### DISCUSSION TITLE
Building the source CLP on MS Visual Studio 2019

### DISCUSSION THREAD

**Discussion by vpranjal** [coin-or/Clp#182] [Q&A] [answered]
Hi

I was building the source code (CLP to start with: downloaded from https://www.coin-or.org/download/source/Clp/ version 1.17.6) on MS Visual Studio 2019 community edition on Windows 10.
![image](https://user-images.githubusercontent.com/59410798/113366489-f9204e00-9376-11eb-9d06-33a4c9b1ba18.png)

It turns out there are some LNK1181 errors when building the Example. 
Interestingly libClp and libCoinUtils build when tried independently (CoinUtils unit test also has the LNK1181 issue in that particular solution). Has anyone faced this issue before? Any ideas!

Thanks

**Comment by jhmgoossens** [comment]
I believe the sources you mention are nearly a year old. Could you try the same with the current? I'm doing some minor cleanup of the VS2019 solution (like #161). I'll also include the example projects.

**Reply by vpranjal** [reply]
Hi

I took the latest Clp from repo as a zip file and tried to work with the .sln file.
The latest .sln file appears to be in MSVisualStudio\v16\libClp

Building this libClp.sln multiple header files missing errors. I think that there are dependencies that prevent the .sln to be built directly?? Let me know if I am missing something here.

![image](https://user-images.githubusercontent.com/59410798/113393628-8df56c80-93b4-11eb-8227-dbfec212719a.png)

Thanks

**Comment by jhmgoossens** [comment, marked as answer]
:-) Yes, all the dependencies. Have a look at https://coin-or.github.io/user_introduction.html
In short, the advised way is to let coinbrew grab all the dependencies for you and do the build. 

Personally, I prefer to [build without coinbrew](https://coin-or.github.io/user_introduction.html#building-without-coinbrew) so git clone (or download as zip) the dependencies manually and build in Visual Studio. For Clp, this means to git clone Clp, CoinUtils, Osi, and BuildTools, and make sure they are in the same root folder (like repos/Clp, repos/CoinUtils, repos/Osi, repos/BuildTools) such that the relative paths in the solution & projects find each other. The VS2019 (v16) solutions and projects should work.

**Reply by vpranjal** [reply]
Thanks- this works- libClp and libOsiClp builds (cloned dependencies manually)

I tried copying the examples and including them in the libClp.sln manually (copied from v10): this didn't work - > This is mostly because the .vcxproj files are on VS2010 and don't work with 2019.

Are there some steps to include the examples from the examples directory and run them in the same solution? Running the cpp files directly from the examples directory expects. (I tried copying the project to v16 folder-> adding example project to solution-> add libClp and libOsiClp that were successfully built independently as dependencies to the example project-> run an example testBarrier.cpp this leads to cannot open multiple .hpp files issues)-> are there any more dependencies to be built?
![image](https://user-images.githubusercontent.com/59410798/113400718-dfefbf80-93bf-11eb-928e-b4624356ed48.png)


Also, I noticed https://github.com/coin-or-tools/BuildTools is on v10? I didn't have to build it but just making sure I have the latest one.

**Reply by vpranjal** [reply]
If I link the ..\..\..\scr into the include additional directories these issues go away from files of the example project. But executing any file from example leads to 
![image](https://user-images.githubusercontent.com/59410798/113405044-ca31c880-93c6-11eb-95ae-5c45661e6776.png)

I do not think this is the right way to add examples to the solution that I m trying here.

![image](https://user-images.githubusercontent.com/59410798/113405207-111fbe00-93c7-11eb-91a3-1a942835bea1.png)

The weird thing is that libClp builds perfectly -> however, when I scope into a file eg here: ClpSimplex.hpp -> it inturn has headers that are not resolved-> I m not sure how it builds and rebuilds libClp successfully even then !!

**Reply by jhmgoossens** [reply]
The Visual Studio files have been somewhat left behind. The VS2019 Clp solution should build without any other action.
The examples projects, indeed, have not been upgraded yet. Should not be too hard. Something for the ToDo list.

**Reply by jhmgoossens** [reply]
Please try again. Now (as of #183) the VS2019/v16 Clp solution has been added together with the v16 version of the mission example projects etc.

**Reply by vpranjal** [reply]
Thanks for the #183.
The build works now. I am able to run the .exes generated from the solution build and solve the sample data problem.

The build from Visual Studio run leads to 
![image](https://user-images.githubusercontent.com/59410798/113456457-f2531300-942a-11eb-8fe4-c23c13923e91.png)

Is this expected? It works from the built exes. {I will be able to solve the problems I place in Data/Sample/}
I was expecting to build the source and walk through the solver solving it via debugger and understand it.  

Let me know !!

**Reply by jhmgoossens** [reply]
It looks like you are trying to start libClp, but that is not possible because it is a library, not an application/exe itself. The exes _use_ the libClp library, so you can use the example exes do step through etc.

**Reply by vpranjal** [reply]
Thanks for this- It works now- I had to change the startup project to ClpExamples- and things work in debug mode on Visual Studio.
